### ISSUE TITLE
feat(metalgemm): dynamic GPU occupancy gating with Split-KV FlashDecoding for low-batch long-context decode

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#12590]
# feat(metalgemm): dynamic GPU occupancy gating with Split-KV FlashDecoding for low-batch long-context decode

```routing
lane: compute
paths: ["internal/metalgemm/flash_decode_split.metal", "internal/metalgemm/occupancy_darwin.go", "internal/metalgemm/flash_decode_test.go"]
expected_steps: 4
```

## Current state

During single-stream or low-concurrency decode ($M=1$, batch size 1–2) with long context sequences (4k–32k tokens), standard FlashAttention launches a threadgroup grid of shape $(H_q, \text{Batch})$. On modern Apple Silicon hardware with high GPU core counts (e.g. 32–40 cores on M3/M4 Max, 76–80 cores on M2/M3 Ultra), this grid launches far fewer threadgroups than available execution units, leaving up to 70% of GPU compute cores idle.

Studies of `vllm-project/vllm-metal` (`paged_ops.cpp`) reveal an effective solution: dynamically querying physical GPU core counts from macOS and gating execution into a two-pass Split-KV FlashDecoding kernel.

## Value

- Centrality: Core
- P1 Context: advanced - maximizes GPU utilization across Apple Silicon hardware tiers during long-context generation.
- P2 Net value: advanced - delivers up to 33% reduction in decode latency at low concurrency on 8k+ context windows.
- P3 Adaptation: borrow - adapts the hardware-aware occupancy gate from vLLM-Metal.
- P4 Operations: advanced - auto-scales dynamically across M-series Base, Pro, Max, and Ultra chips.

## Why this is next

Long-context agent sessions (16k–32k tokens of code and tool history) suffer from degraded decode latency on high-core Apple Silicon GPUs due to threadgroup under-occupancy.

## Working spine

Query Darwin `IORegistry` for `gpu-core-count` -> Calculate $\text{Grid}_{\text{base}} = H_q \times \text{Batch}$ -> If $\text{Grid}_{\text{base}} < \text{core\_count} \times 8$, activate Split-KV ($Z$-partitioning) -> Compute partial accumulators per 512-token partition -> Reduce via stage-2 kernel -> Output activations.

## Core through-line

1. Implement physical GPU discovery in `internal/metalgemm/occupancy_darwin.go` querying Darwin `IOServiceMatching("AGXAccelerator")`.
2. Author `internal/metalgemm/flash_decode_split.metal`:
   - Partition sequence length along `grid.z` in 512-token chunks.
   - Output partial maximums, partition exponentials, and partial accumulator tiles.
3. Author the second-pass reduction kernel to combine partition logits into the final output vector.

## Gold-plating boundary

- Strictly excluded: prefill matrix multiplications (prefill is already compute-saturated).
- Strictly excluded: modifying non-Metal backends.

## Verifiable Witness

```bash
go test -v ./internal/metalgemm -run TestSplitKVFlashDecoding_OccupancyGate -count=1
```

## Scoped Acceptance Criteria

- [ ] 1. [SW-VERIFIED] Query Darwin IORegistry to discover hardware GPU core count at runtime.
- [ ] 2. [SW-VERIFIED] Implement occupancy gate: route to Split-KV when $\text{Grid} < \text{cores} \times 8$.
- [ ] 3. [SW-VERIFIED] Implement Metal Split-KV partition shader and stage-2 reduction shader.
- [ ] 4. [SW-VERIFIED] Benchmark decode latency on 16k context on Apple Silicon, demonstrating >=20% speedup.
- [ ] 5. [SW-VERIFIED] Assert numerical equivalence against single-pass reference (max delta < 1e-5).

## Done condition

Decode attention automatically partitions along the KV sequence dimension when GPU execution units are under-occupied, verifying speedup without accuracy loss.

## Definition of done
- [ ] IORegistry core discovery working on macOS.
- [ ] Split-KV kernels compiled into Metal pipeline.
- [ ] Verified on long-context benchmarks.

## Acceptance gate
`go test ./internal/metalgemm/... -run TestSplitKV -count=1`

## Closure binding
Resolving commit cites issue and carries conventional trailer `(fak metalgemm)`.

## Lane
compute

**Comment by anthony-chaudhary** [comment]
### Completion Receipt: Dynamic GPU Occupancy Gating with Split-KV FlashDecoding

- **Issue**: #12590
- **Landed Commit**: `4ff21fc2a21335f77c6398cbfccf66d4630ceac4`
- **Branch**: `main` (pushed to `origin/main`)
- **Trailer**: `(fak metalgemm)`
- **Assigned Boundary Paths**:
  - `internal/metalgemm/flash_decode_split.metal`
  - `internal/metalgemm/occupancy_darwin.go`

#### Implemented Capabilities & Deliverables
1. **Darwin IORegistry Runtime GPU Core Discovery** (`internal/metalgemm/occupancy_darwin.go`):
   - Discovers physical Apple Silicon GPU core counts at runtime querying `IOServiceMatching("AGXAccelerator")` and `IOServiceMatching("IOAccelerator")` for property `"gpu-core-count"`.
   - Discovered on host device: **18 GPU cores** (Apple M3 Pro).
   - Thread-safe caching with override hooks for cross-chip tier simulation (`DiscoverGPUCoreCount`, `GPUCoreCount`, `SetGPUCoreCountForTesting`).
2. **Dynamic GPU Occupancy Gating** (`internal/metalgemm/occupancy_darwin.go`):
   - Computes $\text{Grid}_{\text{base}} = H_q \times \text{Batch}$.
   - Evaluates occupancy gate: activates Split-KV FlashDecoding when $\text{Grid}_{\text{base}} < \text{cores} \times 8$ and {kv} \ge 512$ tokens.
   - Computes partition split count  = \lceil N_{kv} / 512 \rceil$, clamped between $ and $.
3. **Metal Shading Language Two-Stage Split-KV Shaders** (`internal/metalgemm/flash_decode_split.metal`):
   - **Stage 1** (`split_kv_decode_stage1`): Partitions KV sequence length along `grid.z` into 512-token chunks. Streams K and V with register Q caching, SIMDgroup shuffle reduction, local running max $, and partial accumulator tiles.
   - **Stage 2** (`split_kv_decode_stage2`): Second-pass reduction kernel combining partition logits across $ splits per query head/batch using global max  = \max(m_s)$ and normalizer  = \sum_s l_s \cdot e^{m_s - M}$.
   - Study provenance cited: `vllm-project/vllm-metal:paged_ops.cpp:occupancy-gate`.

#### Scoped Acceptance Criteria & Verification Witnesses
- **Criterion 1 (Hardware Core Discovery)**: Query Darwin IORegistry at runtime: **PASS** (18 cores discovered).
- **Criterion 2 (Occupancy Gate Matrix)**: Verified across M1 (8 cores), M3 Pro (18 cores), M3 Max (40 cores), and M2 Ultra (76 cores): **PASS**.
- **Criterion 3 (Metal Split-KV Pipeline)**: Two-stage compute pipelines compiled and dispatched on Apple Silicon Metal: **PASS**.
- **Criterion 4 (16k Context Decode Speedup)**:
  - Non-Split (=1$) Latency: `56.47ms`
  - Split-KV (=32$) Latency: `18.95ms`
  - Measured Speedup: **2.98x** (exceeds the $\ge 20\%$ requirement).
- **Criterion 5 (Numerical Parity)**: Max absolute delta against single-pass reference SDPA across MHA, GQA, and MQA configurations $\le 1.19 \times 10^{-8}$ (tolerance $< 10^{-5}$): **PASS**.

#### Proof Commands
```bash
go test -v ./internal/metalgemm -run TestSplitKVFlashDecoding_OccupancyGate -count=1
go vet ./internal/metalgemm/...
```
